### PR TITLE
Improve tests, fix unicode conversion to latin1

### PR DESCRIPTION
**Pull request by fulior**
filter-out activities that are already sent

### CODE REVIEW

File: ckanext/dataset_subscriptions/actions.py [hunk 1]
```diff
@@ -39,8 +42,13 @@ def dms_notification_provider(user_dict, since):
                      if activity['user_id'] != user_dict['id']
                      and 'package' in activity['activity_type']]
     # We want a notification per changed dataset, not a list of all changes
-    deduplicated_activity_list = list({item["object_id"]: item for item in activity_list}.values())
+    deduplicated_activity_list = list({item["object_id"]: item for item in sorted(activity_list, key = lambda item: item['timestamp'])}.values())
```
Comment (tomeksabala) [inline ckanext/dataset_subscriptions/actions.py]: too much happening in this line at once, can you split the logic into multiple lines for better readability?
Maybe:
1. sort by timestamp
2. remove duplicates?

File: ckanext/dataset_subscriptions/actions.py [hunk 2]
```diff
@@ -39,8 +42,13 @@ def dms_notification_provider(user_dict, since):
...
     updated_activity_list = _add_dataset_name_to_activity_list(deduplicated_activity_list, context)
+    # Filter out the old activities.
+    strptime = datetime.datetime.strptime
+    fmt = '%Y-%m-%dT%H:%M:%S.%f'
+    updated_activity_list = [activity for activity in updated_activity_list
+            if strptime(activity['timestamp'], fmt) > since]
```
Comment (tomeksabala) [inline ckanext/dataset_subscriptions/actions.py]: this piece should be a separate function. Then you can refactor the comment into a function callout.

File: ckanext/dataset_subscriptions/tests/test_actions.py [hunk 3]
```diff
@@ -44,22 +46,35 @@ def test_with_no_dataset_updates():
...
     assert [ "2 new activities from CKAN" in notification["subject"]  for notification
             in notifications ]
+
+
+@pytest.mark.usefixtures("with_request_context","clean_db")
+@pytest.mark.ckan_config("ckan.activity_streams_email_notifications", True)
+@pytest.mark.usefixtures("with_plugins")
+@pytest.mark.usefixtures("mail_server")
+@mock.patch("ckan.logic.action.update.request")
+def test_send_notification(mock_request):
```
Comment (tomeksabala) [inline ckanext/dataset_subscriptions/tests/test_actions.py]: what does this test exactly do? Can you provide more descriptive name for it?

File: ckanext/dataset_subscriptions/tests/test_actions.py [hunk 4]
```diff
@@ -44,22 +46,35 @@ def test_with_no_dataset_updates():
...
     notifications = email_notifications.get_notifications(user, datetime.datetime.fromtimestamp(3600))
     # we expect two notifications - one for a new dataset and one for a resource
     assert [ "2 new activities from CKAN" in notification["subject"]  for notification
             in notifications ]
+
+
+@pytest.mark.usefixtures("with_request_context","clean_db")
+@pytest.mark.ckan_config("ckan.activity_streams_email_notifications", True)
+@pytest.mark.usefixtures("with_plugins")
+@pytest.mark.usefixtures("mail_server")
```
Comment (tomeksabala) [inline ckanext/dataset_subscriptions/tests/test_actions.py]: A single use of `@pytest.mark.usefixtures("fixture1", "fixture2", ...)` would look here much cleaner. Can you fix all of the annotations so we have consistency withing the file?

**Review by tomeksabala — CHANGES_REQUESTED** [review]

File: ckanext/dataset_subscriptions/tests/test_actions.py (hunk 4 above)
Comment (fulior) [reply ckanext/dataset_subscriptions/tests/test_actions.py]: Would it? I think separate instances of usefixtures for key functionality is easier to read than one line with multiple fixtures. I did it on purpose here, as "emailasusername" extension does the same"

File: ckanext/dataset_subscriptions/tests/test_actions.py (hunk 3 above)
Comment (fulior) [reply ckanext/dataset_subscriptions/tests/test_actions.py]: it's testing "send_notification" function. Should I rename it to "test_send_notification_function" then?

File: ckanext/dataset_subscriptions/actions.py (hunk 1 above)
Comment (fulior) [reply ckanext/dataset_subscriptions/actions.py]: done

File: ckanext/dataset_subscriptions/actions.py (hunk 2 above)
Comment (fulior) [reply ckanext/dataset_subscriptions/actions.py]: done

File: ckanext/dataset_subscriptions/tests/test_actions.py (hunk 3 above)
Comment (fulior) [reply ckanext/dataset_subscriptions/tests/test_actions.py]: done

**Review by tomeksabala — APPROVED** [review]

**Pull request merged**